src/browser/views.py

```diff
@@ -373,6 +374,277 @@ def repo_settings(request, repo_base, repo):
+        for table in base_tables:
+            # check if license view exists for this license_id
+            applied = manager.check_license_applied(table, repo, license_id)
+            license_applied.append(applied)
+
+    license = LicenseManager.find_license_by_id(license_id)
+
+    table_info_tuples = [(x, y) for x in base_tables for y in license_applied]
+
+    rls_table = 'policy'
+
+    res = {
+        'license': license,
+        'login': username,
+        'repo_base': repo_base,
+        'repo': repo,
+        'table_info_tuples': table_info_tuples,
+        'license_views': license_views,
+        'rls-table': rls_table,
+        'collaboratoes': collaborators,
+        }
+
+    res.update(csrf(request))
+
+    return render_to_response("repo-license-manage.html", res)
+
+
+@csrf_exempt
+@login_required
+def link_license(request, repo_base, repo, license_id):
+    '''
+    links a license with a particular repo
+    '''
+    username = request.user.get_username()
+    public_role = settings.PUBLIC_ROLE
+
+    with DataHubManager(user=username, repo_base=repo_base) as manager:
+        collaborators = manager.list_collaborators(repo)
+
+    # if the public role is in collaborators, note that it's already added
+    repo_is_public = next(
+        (True for c in collaborators if
+            c['username'] == settings.PUBLIC_ROLE), False)
+
+    # remove the current user, public user from the collaborator list
+    # collaborators = [c.get('username') for c in collaborators]
+
+    collaborators = [c for c in collaborators if c['username']
+                     not in ['', username, settings.PUBLIC_ROLE]]
+
+    LicenseManager.create_license_link(repo_base, repo, license_id)
+
+    repo_licenses = LicenseManager.find_licenses_by_repo(repo_base, repo)
+    all_licenses = LicenseManager.find_licenses()
+
+    return HttpResponseRedirect(reverse('browser-repo_licenses',
+                                args=(repo_base, repo)))
+
+
+@csrf_exempt
+@login_required
+def license_create(request):
+    username = request.user.get_username()
+    public_role = settings.PUBLIC_ROLE
+
+    if request.method == 'POST':
+        # creates a new license
+        try:
+            pii_def = None
+            pii_anonymized = False
+            pii_removed = False
+            # project_name = ""
+
+            license_name = request.POST['license_name']
+            # project_name = request.POST['project_name']
+            pii_def = request.POST['pii_def']
+            if 'anonymized' in request.POST.getlist('pii_properties'):
+                pii_anonymized = True
+            if 'removed' in request.POST.getlist('pii_properties'):
+                pii_removed = True
+
+            LicenseManager.create_license(
+                license_name=license_name,
+                pii_def=pii_def,
+                pii_anonymized=pii_anonymized,
+                pii_removed=pii_removed)
+
+        except Exception as e:
```

Comment:
Should this be like, "raise e" instead, or like c = {info about error}; c.update(csrf(request)); return render_to_response("template.html", c") ?

Reply:
I mean I don't think there should be a try/except there at all. Most of it could be written to either never fail or validate inputs. If you search the project for uses of "raise ValueError" you'll see some examples of that. Below is how I would change that block of code.

The only external function call in that POST section is `LicenseManager.create_license()`, and that should raise its own meaningful exceptions. Since it's just a SQL call underneath, that should be fine as is, since there's already code in place to convert Postgres exceptions into meaningful HTTP errors.

```py
    if request.method == 'POST':
        # creates a new license

        # Define fallback to prevent MultiDictKeyError if license_name is undefined.
        license_name = request.POST['license_name'] or None
        # project_name = request.POST['project_name'] or ""
        pii_def = request.POST['pii_def'] or None
        pii_anonymized = 'anonymized' in request.POST.getlist('pii_properties')
        pii_removed = 'removed' in request.POST.getlist('pii_properties')

        # Validate parameters and raise descriptive exceptions as needed.
        if not license_name:
            # An unhandled ValueError is returned as a 400 HTTP status code to the client.
            raise ValueError("Request missing \'license_name\' parameter.")

        LicenseManager.create_license(
            license_name=license_name,
            pii_def=pii_def,
            pii_anonymized=pii_anonymized,
            pii_removed=pii_removed)

        return HttpResponseRedirect(reverse('browser-user', args=(username,)))
```